graphql/src/store/prefetch.rs

```diff
@@ -0,0 +1,752 @@
+enum TypeCondition {
+    Any,
+    On(q::Name),
+}
+
+impl TypeCondition {
+    /// Return a `TypeCondition` that matches when `self` and `other` match
+    /// simultaneously. If the two conditions can never match at the same time,
+    /// return `None`
+    fn and(&self, other: &Self) -> Option<Self> {
+        use TypeCondition::*;
+        match (self, other) {
+            (Any, _) => Some(other.clone()),
+            (_, Any) => Some(self.clone()),
+            (On(name1), On(name2)) if name1 == name2 => Some(self.clone()),
+            _ => None,
+        }
+    }
+
+    fn matches(&self, entities: &Vec<Node>) -> bool {
+        use TypeCondition::*;
+        match self {
+            Any => true,
+            On(name) => entities.iter().any(|entity| entity.typename() == name),
+        }
+    }
+
+    /// Return the type that matches this condition; for `Any`, use `object_type`
+    fn matching_type<'a>(
+        &self,
+        schema: &'a s::Document,
+        object_type: &'a ObjectOrInterface<'a>,
+    ) -> Option<ObjectOrInterface<'a>> {
+        use TypeCondition::*;
+
+        match self {
+            Any => Some(object_type.to_owned()),
+            On(name) => object_or_interface_by_name(schema, name),
+        }
+    }
+}
+
+impl From<Option<q::TypeCondition>> for TypeCondition {
+    fn from(cond: Option<q::TypeCondition>) -> Self {
+        match cond {
+            None => TypeCondition::Any,
+            Some(q::TypeCondition::On(name)) => TypeCondition::On(name),
+        }
+    }
+}
+
+/// Intermediate data structure to hold the results of prefetching entities
+/// and their nested associations. For each association of `entity`, `children`
+/// has an entry mapping the response key to the list of nodes.
+#[derive(Debug, Clone)]
+struct Node {
+    entity: Entity,
+    children: BTreeMap<String, Vec<Rc<Node>>>,
+}
+
+impl From<Entity> for Node {
+    fn from(entity: Entity) -> Self {
+        Node {
+            entity,
+            children: BTreeMap::default(),
+        }
+    }
+}
+
+/// Convert a list of nodes into a `q::Value::List` where each node has also
+/// been converted to a `q::Value`
+fn node_list_as_value(nodes: Vec<Rc<Node>>) -> q::Value {
+    q::Value::List(
+        nodes
+            .into_iter()
+            .map(|node| Rc::try_unwrap(node).unwrap_or_else(|rc| rc.as_ref().clone()))
+            .map(|node| node.into())
+            .collect(),
+    )
+}
+
+/// We pass the root node of the result around as a vec of nodes, not as
+/// a single node so that we can use the same functions on interior node
+/// lists which are the result of querying the database. The root list
+/// consists of exactly one entry, and that entry has an empty
+/// (not even a `__typename`) entity.
+///
+/// That distinguishes it from both the result of a query that matches
+/// nothing (an empty `Vec`), and a result that finds just one entity
+/// (the entity is not completely empty)
```

Comment:
Inserting a `__typename` is part of the deserialization logic in the store; I completely agree that a lot of the code dealing with entities and `q::Value` is _very_ dynamically typed. It would be nice to address that (e.g., by turning `Entity` into a struct that has to have a `__typename`) but that's a much bigger undertaking.